src/query/src/optimizer/string_normalization.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use datafusion::config::ConfigOptions;
+use datafusion::logical_expr::expr::Cast;
+use datafusion_common::tree_node::{Transformed, TreeNode, TreeNodeRewriter};
+use datafusion_common::{Result, ScalarValue};
+use datafusion_expr::{Expr, LogicalPlan};
+use datafusion_optimizer::analyzer::AnalyzerRule;
+
+/// StringNormalizationRule normalizes(trims) string values in logical plan.
+/// Mainly used for timestamp trimming
+pub struct StringNormalizationRule;
+
+impl AnalyzerRule for StringNormalizationRule {
+    fn analyze(&self, plan: LogicalPlan, _config: &ConfigOptions) -> Result<LogicalPlan> {
+        plan.transform(&|plan| {
+            let mut converter = StringNormalizationConverter;
+            let inputs = plan.inputs().into_iter().cloned().collect::<Vec<_>>();
+            let expr = plan
+                .expressions()
+                .into_iter()
+                .map(|e| e.rewrite(&mut converter))
+                .collect::<Result<Vec<_>>>()?;
+            datafusion_expr::utils::from_plan(&plan, &expr, &inputs).map(Transformed::Yes)
+        })
+    }
+
+    fn name(&self) -> &str {
+        "StringNormalizationRule"
+    }
+}
+
+struct StringNormalizationConverter;
+
+impl TreeNodeRewriter for StringNormalizationConverter {
+    type N = Expr;
+
+    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
+        let new_expr = match expr {
+            Expr::Cast(Cast { expr, data_type }) => {
```

Comment:
maybe we need an extra check here that `data_type` is Timestamp

Reply:
Yes please